cmd/influxd/upgrade/upgrade.go

```diff
@@ -478,54 +484,85 @@ func runUpgradeE(cmd *cobra.Command, options *options, verbose bool) error {
 	return nil
 }
 
-// validatePaths ensures that all filesystem paths provided as input
-// are usable by the upgrade command
-func validatePaths(sourceOpts *optionsV1, targetOpts *optionsV2) error {
-	if sourceOpts.dbDir != "" {
-		fi, err := os.Stat(sourceOpts.dbDir)
+// validatePaths ensures that all paths pointing to V1 inputs are usable by the upgrade command.
+func (o *optionsV1) validatePaths() error {
+	if o.dbDir != "" {
+		fi, err := os.Stat(o.dbDir)
 		if err != nil {
-			return fmt.Errorf("1.x DB dir '%s' does not exist", sourceOpts.dbDir)
+			return fmt.Errorf("1.x DB dir '%s' does not exist", o.dbDir)
 		}
 		if !fi.IsDir() {
-			return fmt.Errorf("1.x DB dir '%s' is not a directory", sourceOpts.dbDir)
+			return fmt.Errorf("1.x DB dir '%s' is not a directory", o.dbDir)
 		}
 	}
 
-	metaDb := filepath.Join(sourceOpts.metaDir, "meta.db")
+	metaDb := filepath.Join(o.metaDir, "meta.db")
 	_, err := os.Stat(metaDb)
 	if err != nil {
 		return fmt.Errorf("1.x meta.db '%s' does not exist", metaDb)
 	}
 
-	if targetOpts.configPath != "" {
-		if _, err := os.Stat(targetOpts.configPath); err == nil {
-			return fmt.Errorf("file present at target path for upgraded 2.x config file '%s'", targetOpts.configPath)
+	return nil
+}
+
+// validatePaths ensures that none of the paths pointing to V2 outputs refer to existing files.
+func (o *optionsV2) validatePaths() error {
+	if o.configPath != "" {
+		if _, err := os.Stat(o.configPath); err == nil {
+			return fmt.Errorf("file present at target path for upgraded 2.x config file '%s'", o.configPath)
 		}
 	}
 
-	if _, err = os.Stat(targetOpts.boltPath); err == nil {
-		return fmt.Errorf("file present at target path for upgraded 2.x bolt DB: '%s'", targetOpts.boltPath)
+	if _, err := os.Stat(o.boltPath); err == nil {
+		return fmt.Errorf("file present at target path for upgraded 2.x bolt DB: '%s'", o.boltPath)
 	}
 
-	if fi, err := os.Stat(targetOpts.enginePath); err == nil {
+	if fi, err := os.Stat(o.enginePath); err == nil {
```

Comment:
Do we want an `else` clause here for if the` os.stat()` can't be reached.  I see two cases, an err if the directory doesn't exist, which would be ok, and an err if there is a permissions problem or missing mount or filesystem error, which isn't ok.  Maybe these error cases are handled later in the directory creation process.

Reply:
I don't think it's worth it - for one thing, it's going to be very hard to tell which type of error we hit:

https://golang.org/pkg/os/#Stat

> Stat returns a FileInfo describing the named file. If there is an error, it will be of type *PathError.

Different OS's are going to handle mount failures etc. different. If you're trying to upgrade at an existing path while the drive is mounting you probably deserve problems.

Reply:
I'm thinking we catch the Stat() error, and then just call os.IsNotExist(err error) bool. If it's true, we are happy, if it is false, we report the error.

Reply:
There is an `os.IsExist(err error)` function, I could use it instead of `err == nil` in all of these conditional checks.

Reply:
Scratch that, looks like `os.IsExist` is checking for a specific error and not a lazy shorthand for `!os.IsNotExist` like I was assuming.

Reply:
@davidby-influx sanity-checking: your comment isn't really specific to the engine directory, right? We could say the same for all the other `Stat` calls in this function.

Reply:
And then return or handle the error if os.IsExist() == false.  That would let you catch the edge cases where you get an error you don't expect and want to surface it.

Reply:
Yes, my point (and perhaps we are spending too much time on this small point) is that we want to distinguish the case the code handles, the presence of a file where there should not be one), from the case where we get some file system error trying to check of that existence.